### ISSUE TITLE
🐛 Bug Report: Backstage 1.24. breaks scaffolder 

### ISSUE THREAD

**Issue by dariozachow** [backstage/backstage#23760]
### 📜 Description

I've upgraded backstage to version 1.24 and if I try to execute a template 401 unauthorized is displayed.

### 👍 Expected behavior

It should run the template as it was in previous versions.

### 👎 Actual Behavior with Screenshots

When clicking on create the scaffolder shows an error 401 Unauthorized. No Description or anything.

![image](https://github.com/backstage/backstage/assets/147033205/c27ea555-d7c7-4f98-9f48-a7c62e4b930c)

### 👟 Reproduction steps

Go to Scaffolder and run any template.

### 📃 Provide the context for the Bug.

I think this occurence of the bug comes from the update in the auth system of backstage.
After debugging a little bit, I could not find the exact location where the unauthorized request is made.
But adding the following code to permissions plugin fixes this error.

```ts
http.addAuthPolicy({
    path: "/",
    allow: "unauthenticated"
 });
```

I think the request must be made by the scaffolder-backend-plugin somewhere, as no errors are shown in the network tab.

### 🖥️ Your Environment

OS:   Darwin 23.4.0 - darwin/arm64
node: v18.17.1
yarn: 3.7.0
cli:  0.26.0 (installed)
backstage:  1.24.0

Dependencies:
  @backstage/app-defaults                                          1.5.1
  @backstage/backend-app-api                                       0.5.14, 0.6.0
  @backstage/backend-common                                        0.19.10, 0.21.4
  @backstage/backend-defaults                                      0.2.14
  @backstage/backend-dev-utils                                     0.1.4
  @backstage/backend-openapi-utils                                 0.1.7
  @backstage/backend-plugin-api                                    0.6.14
  @backstage/backend-tasks                                         0.5.19
  @backstage/backend-test-utils                                    0.3.4
  @backstage/catalog-client                                        1.6.1
  @backstage/catalog-model                                         1.4.5
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.4
  @backstage/cli                                                   0.26.0
  @backstage/config-loader                                         1.7.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.12.1
  @backstage/core-compat-api                                       0.2.1
  @backstage/core-components                                       0.13.10, 0.14.1
  @backstage/core-plugin-api                                       1.9.1
  @backstage/dev-utils                                             1.0.28
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.6
  @backstage/frontend-plugin-api                                   0.6.1
  @backstage/integration-aws-node                                  0.1.10
  @backstage/integration-react                                     1.1.25
  @backstage/integration                                           1.9.1
  @backstage/plugin-api-docs                                       0.11.1
  @backstage/plugin-app-backend                                    0.3.62
  @backstage/plugin-app-node                                       0.1.14
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.6
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.5
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.9
  @backstage/plugin-auth-backend-module-github-provider            0.1.11
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.11
  @backstage/plugin-auth-backend-module-google-provider            0.1.11
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.9
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.11
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.7
  @backstage/plugin-auth-backend-module-oidc-provider              0.1.4
  @backstage/plugin-auth-backend-module-okta-provider              0.0.7
  @backstage/plugin-auth-backend                                   0.22.0
  @backstage/plugin-auth-node                                      0.4.9
  @backstage/plugin-auth-react                                     0.0.1
  @backstage/plugin-catalog-backend-module-gitlab                  0.3.11
  @backstage/plugin-catalog-backend-module-openapi                 0.1.31
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.11
  @backstage/plugin-catalog-backend-module-unprocessed             0.4.0
  @backstage/plugin-catalog-backend                                1.18.0
  @backstage/plugin-catalog-common                                 1.0.22
  @backstage/plugin-catalog-graph                                  0.4.1
  @backstage/plugin-catalog-import                                 0.10.7
  @backstage/plugin-catalog-node                                   1.8.0
  @backstage/plugin-catalog-react                                  1.11.0
  @backstage/plugin-catalog-unprocessed-entities-common            0.0.1
  @backstage/plugin-catalog                                        1.18.0
  @backstage/plugin-events-backend-module-gitlab                   0.2.0
  @backstage/plugin-events-backend-test-utils                      0.1.24
  @backstage/plugin-events-backend                                 0.3.0
  @backstage/plugin-events-node                                    0.3.0
  @backstage/plugin-explore-common                                 0.0.2
  @backstage/plugin-github-actions                                 0.6.12
  @backstage/plugin-home-react                                     0.1.9
  @backstage/plugin-home                                           0.7.0
  @backstage/plugin-kubernetes-backend                             0.16.0
  @backstage/plugin-kubernetes-common                              0.7.5
  @backstage/plugin-kubernetes-node                                0.1.8
  @backstage/plugin-kubernetes-react                               0.3.1
  @backstage/plugin-kubernetes                                     0.11.6
  @backstage/plugin-org                                            0.6.21
  @backstage/plugin-permission-backend-module-allow-all-policy     0.1.11
  @backstage/plugin-permission-backend                             0.5.37
  @backstage/plugin-permission-common                              0.7.13
  @backstage/plugin-permission-node                                0.7.25
  @backstage/plugin-permission-react                               0.4.21
  @backstage/plugin-proxy-backend                                  0.4.12
  @backstage/plugin-scaffolder-backend-module-azure                0.1.6
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.4
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.4
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.4
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.6
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.4
  @backstage/plugin-scaffolder-backend-module-github               0.2.4
  @backstage/plugin-scaffolder-backend-module-gitlab               0.3.0
  @backstage/plugin-scaffolder-backend                             1.22.0
  @backstage/plugin-scaffolder-common                              1.5.1
  @backstage/plugin-scaffolder-node-test-utils                     0.1.0
  @backstage/plugin-scaffolder-node                                0.4.0
  @backstage/plugin-scaffolder-react                               1.8.1
  @backstage/plugin-scaffolder                                     1.19.0
  @backstage/plugin-search-backend-module-catalog                  0.1.18
  @backstage/plugin-search-backend-module-explore                  0.1.18
  @backstage/plugin-search-backend-module-pg                       0.5.23
  @backstage/plugin-search-backend-module-techdocs                 0.1.18
  @backstage/plugin-search-backend-node                            1.2.18
  @backstage/plugin-search-backend                                 1.5.4
  @backstage/plugin-search-common                                  1.2.11
  @backstage/plugin-search-react                                   1.7.7
  @backstage/plugin-search                                         1.4.7
  @backstage/plugin-techdocs-backend                               1.10.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.6
  @backstage/plugin-techdocs-node                                  1.12.0
  @backstage/plugin-techdocs-react                                 1.2.0
  @backstage/plugin-techdocs                                       1.10.1
  @backstage/plugin-user-settings                                  0.8.2
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.1
  @backstage/theme                                                 0.4.4, 0.5.2
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Hi @dariozachow, can you confirm that you are using the new backend system? I'm assuming yes but want to be sure. Also what auth provider are you using?

**Comment by dariozachow** [comment]
Hi, yes I'm using the new backend system and adding the scaffolder-backend the following way:

```ts
const backend = createBackend();
backend.add(import('@backstage/plugin-scaffolder-backend/alpha'));
backend.start();
```

The auth provider is the gitlab auth provider. And I already updated the auth plugin to use the new auth system.

**Comment by awanlin** [comment]
Thanks for following up @dariozachow, will play around with this.

In the mean time can you try: https://backstage.io/docs/tutorials/auth-service-migration#disabling-the-default-auth-policy. Not saying that's the long term solution just want to see if that can be a possible temporary workaround.

**Comment by dariozachow** [comment]
Thanks for your advice and looking into this. 
I've already created a yarn patch, which disables authentication for every request on the permissions API-Endpoint. This works for now.

**Comment by awanlin** [comment]
Ok, I missed that bit about permissions, can you share your policy and details about your setup, please?

**Comment by dariozachow** [comment]
Sure here is our permission setup file: 

```ts
export const permissionModuleCustomPolicy = createBackendModule({
  moduleId: 'custom-policy',
  pluginId: 'permission',
  register(reg) {
    class GitlabGroupPolicy implements PermissionPolicy {
      config: RootConfigService;
      constructor(config: RootConfigService) {
        this.config = config;
      }
      async handle(
        request: PolicyQuery,
        user?: BackstageIdentityResponse,
      ): Promise<PolicyDecision> {
// lot of permissions mapping
      }
    }

    reg.registerInit({
      deps: { policy: policyExtensionPoint, config: coreServices.rootConfig },
      async init({ policy, config }) {
        policy.setPolicy(new GitlabGroupPolicy(config));
      },
    });
  },
});
```

index.ts: 

```ts
import { permissionModuleCustomPolicy } from './plugins/permission';

const backend = createBackend();

backend.add(permissionModuleCustomPolicy());

backend.start();
```

If you need more details about the actual permissions, tell me :)

**Comment by fstoerkle** [comment]
We encounter the same error. After upgrading to 1.24 and moving to Backstage's built-in authentication, software templates fail with the exact same error as reported above. There is no info in the logs, nor any more output in the UI.

We also use a custom policy, however I see in the network console that the `/api/permission/authorize` comes back with 200 and `{ "id": "...", "result": "ALLOW" }`. Furthermore our policy only restricts specific actions (mostly related to catalog), but none related to the scaffolder. Any unhandled permission is allowed by default. Therefore I doubt that the permission handling is related to this issue.

It's the same for all templates, it does not even start the first step of the template but immediately errors with the 401.

**Comment by freben** [comment]
Just released a patch #23797 that fixes one aspect of auth problems - but that particular one probably only helps if you were also using the `unprocessed` catalog module.

**Comment by fstoerkle** [comment]
@freben thanks, just upgraded to 1.24.1, the issue still persists for us (as expected, as we're not using the `unprocessed` module)

**Comment by vinzscam** [comment]
are you importing

```
backend.add(
  import('@backstage/plugin-permission-backend-module-allow-all-policy'),
);
```

in your `backend/src/index.ts`?
If yes, that might be causing the issue

**Comment by dariozachow** [comment]
Just upgraded to 1.24.2 and the issue is gone for me.

**Comment by fstoerkle** [comment]
Upgraded to 1.24.2, issue still persists.

If it makes a difference, our permission policy has this structure:

```ts
async handle(
  request: PolicyQuery,
  user?: BackstageIdentityResponse,
): Promise<PolicyDecision> {
  if (!user) {
    // we require authentication, so we always need a valid user here
    return { result: AuthorizeResult.DENY };
  }

  if (isPermission(...)) {
    ...
    return { result: AuthorizeResult.DENY };
  }

  if (isPermission(...)) {
    ...
    return { result: AuthorizeResult.DENY };
  }

  return { result: AuthorizeResult.ALLOW };
}
```

**Comment by vinzscam** [comment]
thank you @fstoerkle 🙏, I'm looking into it

**Comment by vinzscam** [comment]
I think this issue isn't related to permissions at all. I still can't reproduce the 401 @fstoerkle

**Comment by dariozachow** [comment]
Yes this is 100% no permission policy problem. The Error occurs when requesting the permissions api route. The request is done by the backend-plugin and missing some kind of authorization token, thus giving the error 401.

**Comment by fstoerkle** [comment]
Some more context: we have migrated to the new backend system, including the new auth services (following [these docs](https://backstage.io/docs/tutorials/auth-service-migration)). We require an authenticated user everywhere, and do not allow guest access.

Might be able to debug myself a bit more later today

**Comment by dariozachow** [comment]
Just realised I have been using following feature for the backend:

´´´yaml
backend:
  auth:
    dangerouslyDisableDefaultAuthPolicy: true
´´´

After removing this one the error still persists

**Comment by Parsifal-M** [comment]
Weird, I am on `1.24.0` and don't have this issue in my scaffolder I also have **not** set the `dangerouslyDisableDefaultAuthPolicy` is this on immediately trying to open a template? or after some steps? From the description sounds like as soon as you select the template. :thinking:

**Comment by dariozachow** [comment]
Occurs after clicking on the create button after filling out all input steps, right as the scaffolder tasks starts.

**Comment by freben** [comment]
What was the network call from the browser that returned 401? Was it the `v2/tasks` call?

**Comment by drodil** [comment]
@freben actually there is no 401 network calls, the message just appears out of nowhere. The tasks endpoint even returns the task as `processing` but it's actually failed. No communication at all in the eventstream.

![image](https://github.com/backstage/backstage/assets/1178319/dc2c2edc-2a87-46b1-ad35-5eebcbd01313)

**Comment by dtrexler** [comment]
Hi everyone, I am also experiencing this issue, but I do believe https://github.com/backstage/backstage/pull/23827 resolves it

**Comment by drodil** [comment]
@benjdlambert the following is coming from `/eventstream`:

```
{
    "error": {
        "name": "AuthenticationError",
        "message": "Missing credentials",
        "stack": "AuthenticationError: Missing credentials\n    at DefaultHttpAuthService.credentials (/node_modules/@backstage/backend-app-api/src/services/implementations/httpAuth/httpAuthServiceFactory.ts:150:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
    },
    "request": {
        "method": "GET",
        "url": "/api/scaffolder/v2/tasks/5c5aecd2-58bd-4924-8d1a-8f8ee546065c/eventstream"
    },
    "response": {
        "statusCode": 401
    }
}
```
From backend, I can see one 401 log to authorize endpoint:
```
[0] 2024-03-26T13:40:07.571Z rootHttpRouter info ::1 - - [26/Mar/2024:13:40:07 +0000] "POST /api/permission/authorize HTTP/1.1" 401 397 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" type=incomingRequest
```
Doesn't matter if using guest user or logged in user.

**Comment by fstoerkle** [comment]
if I do another request to `/eventstream`, I get this response back:

```
event: log
data: {"id":1,"taskId":"0b71efc9-c31e-4980-ad6d-8cf85d819024","body":{"message":"Starting up task with 12 steps"},"type":"log","createdAt":"2024-03-26T13:40:56.000Z"}

event: completion
data: {"id":2,"taskId":"0b71efc9-c31e-4980-ad6d-8cf85d819024","body":{"message":"Run completed with status: failed","error":{"name":"ResponseError","message":"Request failed with 401 Unauthorized"}},"type":"completion","createdAt":"2024-03-26T13:40:56.000Z"}
```

**Comment by benjdlambert** [comment]
@drodil, I think that's because you'll need to pass an `Authorization` header to actually see the event stream response. My guess is that once you pass that `Authorization` token you will see something like @fstoerkle

**Comment by Parsifal-M** [comment]
> @drodil, I think that's because you'll need to pass an `Authorization` header to actually see the event stream response. My guess is that once you pass that `Authorization` token you will see something like @fstoerkle

```
event: log
data: {"id":5,"taskId":"292f074f-866f-4a9f-b622-0bd74ae7c3df","body":{"message":"Starting up task with 2 steps"},"type":"log","createdAt":"2024-03-26T13:53:46.183Z"}

event: completion
data: {"id":6,"taskId":"292f074f-866f-4a9f-b622-0bd74ae7c3df","body":{"message":"Run completed with status: failed","error":{"name":"ResponseError","message":"Request failed with 401 Unauthorized"}},"type":"completion","createdAt":"2024-03-26T13:53:46.196Z"}
```

Can confirm :+1:

**Comment by benjdlambert** [comment]
My guess is that https://github.com/backstage/backstage/pull/23827 should fix this. We're gonna be doing a `-next` release today, wondering if it makes sense to try with that `-next` release and then we can ship that as a patch.

Also going to throw together a quick yarn patch to try locally to see if the fix in the PR works which might be a bit quicker than waiting for the `-next` release.

**Comment by drodil** [comment]
Hope to get a patch for this, our artifact store doesn't allow `-next` releases so unless getting a patch, have to wait for the next major :( Also if the #23831 would get included, it would be perfect!

**Comment by freben** [comment]
Those are on the wishlist for a patch yeah

**Comment by benjdlambert** [comment]
I created a simple yarn patch, can @drodil and @fstoerkle see if this works? If you can't get the patch to work with yarn you can go ahead and make the changes manually, thankfully there's not that much.

`.yarn/patches/@backstage-plugin-scaffolder-backend-npm-1.22.2-8edbc9f4f5.patch`
```
diff --git a/dist/cjs/router-B0mT63g8.cjs.js b/dist/cjs/router-B0mT63g8.cjs.js
index 225bea79b696288551a0a91bc52101c8a6655a86..fa407b2275b15c69d7d6854df7f4814f35aff816 100644
--- a/dist/cjs/router-B0mT63g8.cjs.js
+++ b/dist/cjs/router-B0mT63g8.cjs.js
@@ -2677,7 +2677,7 @@ class NunjucksWorkflowRunner {
           user: task.spec.user,
           isDryRun: task.isDryRun,
           signal: task.cancelSignal,
-          getInitiatorCredentials: task.getInitiatorCredentials
+          getInitiatorCredentials: () => task.getInitiatorCredentials()
         });
       }
       for (const tmpDir of tmpDirs) {
@@ -3429,7 +3429,7 @@ async function createRouter(options) {
       secrets: {
         ...req.body.secrets,
         backstageToken: token,
-        initiatorCredentials: JSON.stringify(credentials)
+        __initiatorCredentials: JSON.stringify(credentials)
       }
     });
     res.status(201).json({ id: result.taskId });
```

**Comment by benjdlambert** [comment]
Should be applied on top of `1.24.2`

**Comment by dariozachow** [comment]
Works for me 👍

**Comment by fstoerkle** [comment]
same here, works! awesome 🎉

**Comment by benjdlambert** [comment]
Apologies - we've noticed that there was some issues with the previous patch releases that's causing us some headaches with another patch release, so I think we're going to hold off on both the `-next` release and the `patch` release until tomorrow just to make sure we're in a good state. Sorry for any inconveniences.

**Comment by benjdlambert** [comment]
A new release is hot off the press. Gonna close this issue as it should be fixed in 1.25.0 :tada: Nice little bonus release just in time for Easter holidays. 😅
